### ISSUE TITLE
python3.5-pip is missing bin/pip

### ISSUE THREAD

**Issue by svend** [NixOS/nixpkgs#12190]
I'm using the nixpkgs-unstable channel (nixpkgs-16.03pre74645.1e63074) on Darwin. I installed  python3.5-pip, but the resulting nix package does not contain the pip binary:

```
$ nix-env -i python3.5-pip
installing ‘python3.5-pip-7.1.2’
these paths will be fetched (13.47 MiB download, 87.67 MiB unpacked):
  /nix/store/2qbqm1gnw98x2mbsxl0kq0vgphpgq4i5-python3.5m-setuptools-18.2
  /nix/store/7az2ns2gwlg6jvi60gbn36486ha1hhf5-tcl-8.6.4
  /nix/store/l6gwjripcsa94dh1dcfwd2zr5fn9mh1l-python3-3.5.1
  /nix/store/ln57640j93c869vm6yz8s0h58495jx38-tk-8.6.4
  /nix/store/lqc5rcjlj5w72hwf4hwxb3s70v2dg2vf-python3.5-pip-7.1.2
fetching path ‘/nix/store/7az2ns2gwlg6jvi60gbn36486ha1hhf5-tcl-8.6.4’...

*** Downloading ‘https://cache.nixos.org/nar/0jjvrndnyqx6v3jm7s71hf3hvmc1rz1sxm5dad31s9xl2fx1mfmb.nar.xz’ to ‘/nix/store/7az2ns2gwlg6jvi60gbn36486ha1hhf5-tcl-8.6.4’...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1878k  100 1878k    0     0  1000k      0  0:00:01  0:00:01 --:--:-- 1001k

fetching path ‘/nix/store/ln57640j93c869vm6yz8s0h58495jx38-tk-8.6.4’...

*** Downloading ‘https://cache.nixos.org/nar/1qwds3zg9zqd02rqkyg5dg52panby1ya5k728r5f1hncwx45z6a7.nar.xz’ to ‘/nix/store/ln57640j93c869vm6yz8s0h58495jx38-tk-8.6.4’...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1640k  100 1640k    0     0  1003k      0  0:00:01  0:00:01 --:--:-- 1006k

fetching path ‘/nix/store/l6gwjripcsa94dh1dcfwd2zr5fn9mh1l-python3-3.5.1’...

*** Downloading ‘https://cache.nixos.org/nar/0zp855m8bngcg7qha5l8agv2iwb3j9y8qclhjjwzmh48mpkk2jd0.nar.xz’ to ‘/nix/store/l6gwjripcsa94dh1dcfwd2zr5fn9mh1l-python3-3.5.1’...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 9924k  100 9924k    0     0  1717k      0  0:00:05  0:00:05 --:--:-- 2238k

fetching path ‘/nix/store/2qbqm1gnw98x2mbsxl0kq0vgphpgq4i5-python3.5m-setuptools-18.2’...

*** Downloading ‘https://cache.nixos.org/nar/03d0ijds3c7fppp8xnr8drf2mgj39d7mp109m15yh4jcvl3qm055.nar.xz’ to ‘/nix/store/2qbqm1gnw98x2mbsxl0kq0vgphpgq4i5-python3.5m-setuptools-18.2’...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  346k  100  346k    0     0   358k      0 --:--:-- --:--:-- --:--:--  359k

fetching path ‘/nix/store/lqc5rcjlj5w72hwf4hwxb3s70v2dg2vf-python3.5-pip-7.1.2’...

*** Downloading ‘https://cache.nixos.org/nar/1hx9npigd4bwdkwihpmdmq5ly9jrwfhs0xki527f3y94vbj50ar6.nar.xz’ to ‘/nix/store/lqc5rcjlj5w72hwf4hwxb3s70v2dg2vf-python3.5-pip-7.1.2’...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   336  100   336    0     0    772      0 --:--:-- --:--:-- --:--:--   775

building path(s) ‘/nix/store/7vahkjcq2jk17wpy2pk7g7ln1xpwylfb-user-environment’
created 1624 symlinks in user environment

$ find /nix/store/lqc5rcjlj5w72hwf4hwxb3s70v2dg2vf-python3.5-pip-7.1.2
/nix/store/lqc5rcjlj5w72hwf4hwxb3s70v2dg2vf-python3.5-pip-7.1.2
/nix/store/lqc5rcjlj5w72hwf4hwxb3s70v2dg2vf-python3.5-pip-7.1.2/lib
/nix/store/lqc5rcjlj5w72hwf4hwxb3s70v2dg2vf-python3.5-pip-7.1.2/lib/python3.5
/nix/store/lqc5rcjlj5w72hwf4hwxb3s70v2dg2vf-python3.5-pip-7.1.2/lib/python3.5/site-packages
/nix/store/lqc5rcjlj5w72hwf4hwxb3s70v2dg2vf-python3.5-pip-7.1.2/nix-support
/nix/store/lqc5rcjlj5w72hwf4hwxb3s70v2dg2vf-python3.5-pip-7.1.2/nix-support/propagated-native-build-inputs
```

I see the same problem with python2.7-pip.

**Comment by FRidh** [comment]
1. Why do you need pip? The recommended way to work with Python packages on Nix is by writing expressions to build Python packages.

2. The `pip` package doesn't actually include any script (see `setup.py`, `scripts` is missing from `setup()`.

In `bootstrapped-pip` we manually create the script:
```nix
  installPhase = ''

    # install pip binary
    echo '${python.interpreter} -m pip "$@"' > $out/bin/pip
    chmod +x $out/bin/pip
```
We could do this in the normal `pip` package as well actually.

**Comment by domenkozar** [comment]
Ugh, this is an ugly bug. Because we use bootstrapped-pip to install pip, it doesn't install scripts.

**Comment by FRidh** [comment]
@domenkozar could you explain what you mean? `bootstrapped-pip` is always added as a `buildInput`, but how does that relate to installing scripts?

**Comment by domenkozar** [comment]
@FRidh when `buildPythonPackage` tries to install `pip`, it says already installed, because it finds `bootstrapped-pip`.

**Comment by FRidh** [comment]
We can also just move `bootstrapped-pip` from `buildInputs` to `propagatedBuildInputs`. That way you always have a working pip available.

**Comment by domenkozar** [comment]
I think `bootstrapped-pip` should just become pip. That's the easiest fix.

**Comment by abbradar** [comment]
https://github.com/abbradar/nixpkgs/commit/f0c9e93528298647c025175563664f6eeff06abb and https://github.com/abbradar/nixpkgs/commit/3433053569ee7f8a9a727c985be0c27ea33e8b19 work for me.

**Comment by FRidh** [comment]
My suggestion would work only if you use `buildPythonPackage`, and therefore doesn't work with a Python environment without any packages, so let's not do that.

Making `bootstrapped-pip` simply `pip` as @domenkozar is the easiest since `buildPythonPackage` is circumvented then. 

Adding `installFlags` as @abbradar proposes could maybe be useful for other use cases as well, although I have none in mind now. Unless we have other use cases I don't think we should add an option just yet.

**Comment by abbradar** [comment]
FWIW my flag solution seems most "right" to me (because it's a "small hack" into the otherwise normal pip building as a Python package), but I don't have any strong preference. Not adding `installFlags` argument would lead to excessive copy-paste though (in particular of that call to pip with lots of flags) which I like less.

**Comment by jgeerds** [comment]
I think we should fix this for 16.03

**Comment by domenkozar** [comment]
:+1:
